File: tools/parity/dispatch/git-push-target.ts
// TS dispatch for the git-push-target parity module: maps the shared vector
// function names to the real `src/shared/git-push-target-validation.ts` export.
// `assertGitPushTargetShape` throws on invalid input and returns void on success,
// so we wrap it into a `{ ok, error? }` value to match the Rust `Result` image.

import { assertGitPushTargetShape } from '../../../src/shared/git-push-target-validation'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'assertGitPushTargetShape': {
      const { remoteName, branchName, remoteUrl } = input as {
        remoteName: string
        branchName: string
        remoteUrl?: string
      }
      // Omit remoteUrl when absent so the optional-URL branch is skipped,
      // matching the Rust `None` arm (an explicit `undefined` key would too).
      const target: Record<string, unknown> = { remoteName, branchName }
      if (remoteUrl !== undefined) {
        target.remoteUrl = remoteUrl
      }
      try {
        assertGitPushTargetShape(target)
        return { ok: true }
      } catch (error) {
        return { ok: false, error: (error as Error).message }
      }
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-dispatch/src/modules/git_push_target.rs
//! Parity dispatch for `orca_core::git_push_target` vs
//! `src/shared/git-push-target-validation.ts`.
//!
//! The TS `assertGitPushTargetShape` throws on invalid input and returns void on
//! success; the Rust port returns `Result<(), String>` with the same messages.
//! Both are shaped into `{ ok, error? }` so the JSON images are equal.

use orca_core::git_push_target::validate_git_push_target;
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "assertGitPushTargetShape" => {
            // Vectors carry valid string fields; type-mismatch cases are out of
            // scope because the Rust port validates already-typed inputs.
            let remote_name = input.get("remoteName").and_then(Value::as_str).unwrap_or("");
            let branch_name = input.get("branchName").and_then(Value::as_str).unwrap_or("");
            // Absent remoteUrl -> None, mirroring the omitted-key TS branch.
            let remote_url = input.get("remoteUrl").and_then(Value::as_str);
            match validate_git_push_target(remote_name, branch_name, remote_url) {
                Ok(()) => json!({ "ok": true }),
                Err(message) => json!({ "ok": false, "error": message }),
            }
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
